### PR TITLE
Test Driven Development (Week 5)

### PR DESCRIPTION
**Pull request by prashantiA**
New Features
- Scheduling algorithm with optional attendees
- Tests for the scheduler

Contains changes from #10

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,48 @@
...
+      ArrayList<String> combinedPeople = new ArrayList<String>();
+      combinedPeople.addAll(request.getAttendees());
+      combinedPeople.addAll(request.getOptionalAttendees());
+      Collection<TimeRange> res = queryFromSorted(sortedEvents, new MeetingRequest(combinedPeople, request.getDuration()));
+      if (!res.isEmpty()) {
+        return res;
+      }
+      else if (request.getAttendees().isEmpty()) {
+        return res;
+      }
```
Comment (lee12jin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:37]: You can join these two statements

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,48 @@
...
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<Event> sortedEvents = new ArrayList<Event>(events);
+    Collections.sort(sortedEvents, Event.ORDER_BY_START);
```
Comment (lee12jin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:24]: Awesome good for you for testing the assumption of whether or not the events were sorted!!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,48 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<Event> sortedEvents = new ArrayList<Event>(events);
+    Collections.sort(sortedEvents, Event.ORDER_BY_START); 
+    
+    Collection<String> optional = request.getOptionalAttendees();
+    if (!optional.isEmpty()) {
+      ArrayList<String> combinedPeople = new ArrayList<String>();
+      combinedPeople.addAll(request.getAttendees());
+      combinedPeople.addAll(request.getOptionalAttendees());
+      Collection<TimeRange> res = queryFromSorted(sortedEvents, new MeetingRequest(combinedPeople, request.getDuration()));
```
Comment (lee12jin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:31]: Do the tests pass? I think you need to treat optional attendees differently to have this code pass your newly written tests?

Comment (prashantiA) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:31]: They all pass. The problem basically views it as all or nothing on optional attendees so you can first check with them being mandatory and then check without optional attendees if it fails. 

The one anomaly is with no mandatory attendees and optional attendees that have conflicts all the time (the problem statement was a bit ambiguous in this case).

**Pull request merged**